src/Core/Services/MetadataProviders/MetadataProviderFactory.cs

```diff
@@ -0,0 +1,65 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+using System.IO.Abstractions;
+using System.Net;
+using Azure.DataApiBuilder.Config.ObjectModel;
+using Azure.DataApiBuilder.Core.Configurations;
+using Azure.DataApiBuilder.Core.Resolvers.Factories;
+using Azure.DataApiBuilder.Service.Exceptions;
+using Microsoft.Extensions.Logging;
+
+namespace Azure.DataApiBuilder.Core.Services.MetadataProviders
+{
+    /// <inheritdoc />
+    public class MetadataProviderFactory : IMetadataProviderFactory
+    {
+        private readonly IDictionary<string, ISqlMetadataProvider> _metadataProviders;
+
+        public MetadataProviderFactory(RuntimeConfigProvider runtimeConfigProvider, IQueryManagerFactory engineFactory, ILogger<ISqlMetadataProvider> logger, IFileSystem fileSystem)
+        {
+            _metadataProviders = new Dictionary<string, ISqlMetadataProvider>();
+            foreach ((string dataSourceName, DataSource dataSource) in runtimeConfigProvider.GetConfig().GetDataSourceNamesToDataSourcesIterator())
+            {
+                ISqlMetadataProvider metadataProvider = dataSource.DatabaseType switch
+                {
+                    DatabaseType.CosmosDB_NoSQL => new CosmosSqlMetadataProvider(runtimeConfigProvider, fileSystem),
+                    DatabaseType.MSSQL => new MsSqlMetadataProvider(runtimeConfigProvider, engineFactory, logger, dataSourceName),
+                    DatabaseType.PostgreSQL => new PostgreSqlMetadataProvider(runtimeConfigProvider, engineFactory, logger, dataSourceName),
+                    DatabaseType.MySQL => new MySqlMetadataProvider(runtimeConfigProvider, engineFactory, logger, dataSourceName),
+                    _ => throw new NotSupportedException(dataSource.DatabaseTypeNotSupportedMessage),
+                };
+
+                _metadataProviders.TryAdd(dataSourceName, metadataProvider);
+            }
+        }
+
+        /// <inheritdoc />
+        public ISqlMetadataProvider GetMetadataProvider(string dataSourceName)
+        {
+            if (!_metadataProviders.ContainsKey(dataSourceName))
+            {
+                throw new DataApiBuilderException($"{nameof(dataSourceName)}:{dataSourceName} could not be found within the config", HttpStatusCode.BadRequest, DataApiBuilderException.SubStatusCodes.DataSourceNotFound);
```

Comment:
nit: split long line to have different params on different lines.

Reply:
fixed.